internal/ethapi/api.go

```diff
@@ -973,20 +973,28 @@ func (api *BlockChainAPI) SimulateV1(ctx context.Context, opts simOpts, blockNrO
 // successfully at block `blockNrOrHash`. It returns error if the transaction would revert, or if
 // there are unexpected failures. The gas limit is capped by both `args.Gas` (if non-nil &
 // non-zero) and `gasCap` (if non-zero).
-func DoEstimateGas(ctx context.Context, b Backend, args TransactionArgs, blockNrOrHash rpc.BlockNumberOrHash, overrides *StateOverride, gasCap uint64) (hexutil.Uint64, error) {
+func DoEstimateGas(ctx context.Context, b Backend, args TransactionArgs, blockNrOrHash rpc.BlockNumberOrHash, overrides *StateOverride, blockOverrides *BlockOverrides, gasCap uint64) (hexutil.Uint64, error) {
 	// Retrieve the base state and mutate it with any overrides
 	state, header, err := b.StateAndHeaderByNumberOrHash(ctx, blockNrOrHash)
 	if state == nil || err != nil {
 		return 0, err
 	}
-	if err := overrides.Apply(state, nil); err != nil {
+	chainContext := NewChainContext(ctx, b)
+	blockCtx := core.NewEVMBlockContext(header, chainContext, nil)
+	if blockOverrides != nil {
+		blockOverrides.Apply(&blockCtx)
+	}
+
+	rules := b.ChainConfig().Rules(blockCtx.BlockNumber, blockCtx.Random != nil, blockCtx.Time)
+	precompiles := maps.Clone(vm.ActivePrecompiledContracts(rules))
+	if err := overrides.Apply(state, precompiles); err != nil {
 		return 0, err
 	}
 	// Construct the gas estimator option from the user input
 	opts := &gasestimator.Options{
 		Config:     b.ChainConfig(),
-		Chain:      NewChainContext(ctx, b),
-		Header:     header,
+		Chain:      chainContext,
+		Header:     blockOverrides.MakeHeader(header),
```

Comment:
Please revert this line. The EVM doesn't have access to the header directly anyway. It only accesses block fields via the block context object.

Reply:
Yeah it took me a while to find this, I was taking most of the code from`DoCall` path which doesn't do this. But the problem with the estimate path is that it uses the header in https://github.com/ethereum/go-ethereum/blob/master/eth/gasestimator/gasestimator.go#L221 

Without this line the block overrides are not available in solidity code when running the unit test; this could be because is the test isn't setup correctly or it could genuinely be needed.

Reply:
Riighhtt this made me realize we should do this differently. Right now you're effectively only applying the block overrides to mutate the header and pass that to the gasestimator.

IMO we should pass in the overrides to gasestimator and apply them there just before execution. But that will create an import cycle. Hmm...

Reply:
I was surprised by the number of differences between the two endpoints. They are basically the same thing, except when they are not!

Reply:
Another option is to move the gasestimator package up into the ethapi package to prevent this.

let me know what you think ... 
https://github.com/ethereum/go-ethereum/pull/30695/commits/c5af1757927e2299f7b18f8782a521473029c260

Reply:
This was a deliberate change to move gas estimator out of rpc: https://github.com/ethereum/go-ethereum/pull/28600. let me think about it

Reply:
Alternatively, I could pass it in on the options, but that didn't feel right.